content/en/docs/allow-port-80.md

```diff
@@ -0,0 +1,42 @@
+---
+title: Best Practice - Keep Port 80 Open
+permalink: /docs/allow-port-80
+top_graphic: 1
+date: 2019-01-24
+lastmod: 2019-01-24
+---
+
+{{< lastmod >}}
+
+We occasionally get reports from people who have trouble using the
+HTTP-01 challenge type because they've firewalled off port 80 to their
+web server. Our recommendation is that all servers meant for general web
+use should offer both HTTP on port 80 and HTTPS on port 443. They should
+also send redirects for all port 80 requests, and possibly an HSTS header.
+
+Allowing port 80 doesn't introduce a larger attack surface, because
+requests on port 80 are generally served by the same software that runs
+on port 443.
+
+Closing port 80 doesn't reduce the risk to a person who accidentally
+visits your website via HTTP. In normal circumstances, that person
+would receive a redirect to HTTPS, and their subsequent traffic will be
+protected. If that person was subject to an active MITM, the MITM would
+answer on port 80, so your site would never have a chance to answer
+"connection refused."
+
+Lastly, keeping port 80 open in order to serve a redirect helps get
+people to the right version of your site (the HTTPS version). There are
+various situations beyond your control that might briefly land someone
+on the HTTP version of your site - for instance, automatic linkification
+in emails, or manually typing a domain name. It's better for them to get
+a redirect than an error.
+
```

Comment:
> One thing that you should do, however, is find old links that point to the HTTP version of your site—especially form submission targets, like `<form action="http://www.example.com/form-handler" method="post">`—and update them or encourage others to update them to the HTTPS versions. This reduces the risk that browsers will try to submit sensitive information via HTTP.

Reply:
This seems like a particularly niche case - I'm not sure it's worth including here, though I do appreciate you pointing out that there's an edge case involved.

Reply:
If you are on a network under attack by [sslstrip](https://moxie.org/software/sslstrip/) for example, having the port 80 open is enough to downgrade all communication to plain text. Just drop any "30x redirect", play the HTTP request on the HTTPS port, response to the client with modified HTML, replacing all https:// URL to http:// ones, drop HTST headers if any or remove the secured cookie flag, and so the client always remains on plain text on each request.
Closing the 80 port rises the sslstrip difficulty (full & active lvl4 TCP MITM and not just lvl7 with tiny on-the-fly sed-like modification).

Reply:
By the way, since the goal of this page is specifically to dispel that particular misconception, do you have any suggestions for how I could word it to make things more clear?

Reply:
"They can answer", but it's more difficult to do (lvl3 TCP vs lvl7 HTTP/HTML).

In my opinion, this is a terrible Let's Encrypt error to **forbid** HTTP-01 443 port for people who want to have only this port open.
There is no decent reason to do that, the only given one is *dumb admin not able to configure properly catchall/fallback* (but there is the same problem with 80 port).
At the opposite, there is quite good reason to close the port 80, even it is not the ultimate solution, it avoids some dumb attack. And it will also be the future, 80 port is on the decline.
The only real trouble with 80 port closing is the way browsers handle naked fqdn, and so without HSTS preload, hitting `example.org` will load `http://example.org` and hangs forever until the timeout. But I think HTTPS may be the default for browser in a not too far future. And with HSTS preload, no trouble at all.

For (quite a lot) people, HTTP-01 is the only way to do ACME validation.
  * DNS-01 supposes API to automate DNS handling and no all DNS provided has one and if you self-host your DNS zone, you're doomed without custom software development
  * DNS-01 not usable with DNSSec (which supposes strong isolation between things able to modify DNS and things able to modify content) and worse DANE (strong isolation between DNS modification and cert modification),
  * TLS-ALPN-01 supposes custom HTTPd to response ACME challenge, which is currently not widely available (apache support is experimental, no nginx support, not available on most of SAAS provider…)
  * To avoid the previous point, TLS-ALPN-01 supposes ACME proxy in front of real reverse-proxy or content provider, and those proxies currently support less load and feature than a standard nginx or haproxy.

Considering the fact : 
- 443 is currently (or at least should be, or will be in a near future) the default everywhere
- there is no reason (or at least no reason not relevant to port 80 too) to not support HTTP-01 on 443 only port
-  HTTP-01 is the only massively & simply available ACME challenge, because no strong requirements
Let's Encrypt should support 443 only HTTP-01 and let people deciding if they want to use 80 or 443, like simpleHTTP challenge at the first time.

Reply:
Another features of HTTP-01 not possible on TLS-ALPN-01 :
- Ability to redirect ACME challenge to any URL you want with a `302 redirect` on `/.well-known/acme-challenge/`, in particular to a central PKI authority able to dispatch the obtained certificate to any content endpoint. TLS-ALPN-01 is restricted by design to "endpoint answering the challenge = the one retrieving the certificate".
- No overhead on non-challenge request. On the contrary, TLS-ALPN-01 must analyze all requests (challenge and content) to know if it's a challenge one or not.

Reply:
Another danger about port 80 open or default HTTP on naked FQDN on browser :
- Connect to your bank site, on HTTPS
- Your bank site set a cookie, generally not flagged `secure`
- Input the naked FQDN of your bank site on your browser.
- Your browser connect to it on 80 port, leak the auth cookie in plain text, then redirect to HTTPS… :fearful: 

If no HTTP port open, no TCP connection, no HTTP request, no cookie sent in plain text

Reply:
> Another danger about port 80 open or default HTTP on naked FQDN on browser :
> 
>   * Connect to your bank site, on HTTPS
>   * Your bank site set a cookie, generally not flagged `secure`
>   * Input the naked FQDN of your bank site on your browser.
>   * Your browser connect to it on 80 port, leak the auth cookie in plain text, then redirect to HTTPS… 😨
> 
> If no HTTP port open, no TCP connection, no HTTP request, no cookie sent in plain text
> (And there is always [a lot of banks without HSTS](https://imirhil.fr/tls/banks.html) at least in France…)

If there is an attacker in your network (which could see the auth cookie), he simply can answer port 80 even if the original webserver doesn't, so you have the same problem.

When a (secure) website sets a cookie, it should generally set the `secure` flag to prevent browsers from sending it over unencrypted HTTP connections, or it should use HSTS (ideally with preloading) to prevent browsers from using plain HTTP connections.

However, even when using the `secure` flag, there is still the problem of Session Fixation: An attacker can get the user to visit a website over HTTP where the attacker answers port 80 and injects a Session Cookie that he has previously retrieved from the website (it can even have the `secure` flag and is still accepted by the browser), and then redirects the user to the `https` site (so it looks secure to the user). But when the user logs in, it still has the attacker's session cookie, so the attacker then also has access to the user's account.
To prevent this, the website should set a new session cookie when logging in, or use HSTS to ensure browsers cannot accept cookies from plain HTTP connections.

Reply:
> If there is an attacker in your network (which could see the auth cookie), he simply can answer port 80 even if the original webserver doesn't (as mentioned by @jsha), so you have the same problem.

Active attacker is not the problem here. The cookie is sent in plain text, so **anybody** on the link, or on the side of the link, can just catch it. From my ISP to any IX on the path, and allowing SIGINT with passive sniffing. See Snowden leaks…

> When a (secure) website sets a cookie, it should generally set the secure flag to prevent browsers from sending it over unencrypted HTTP connections, or it should use HSTS (ideally with preloading) to prevent browsers from using plain HTTP connections.

```
 $ curl -qsIL https://www.net444.caisse-epargne.fr/ | rg "(Cookie|Strict)"
Set-Cookie: CEPARAL=2626332864.56399.0000; path=/; Httponly
Set-Cookie: ASP.NET_SessionId=zlbgqjjyrpoavp45tcmdze55; path=/; HttpOnly
Set-Cookie: CEPARAL=2693441728.56399.0000; path=/; Httponly
```
Not secured, and no HSTS… And this is a **BANK**… :fearful:

Reply:
To summarize, we are agree that having port 80 closed still allow **active** attack (but nobody can really protect from them), but closing the 80 port forbid some **passive/dumb** attack and prevent dumb user to self-foot shoot.
Then, arguing you don't have to close your 80 port because you don't protect yourself from active attack is bullshit or pre-Snowden era thinking.
Allowing port 80 is **not** good compared to closing it to make impossible passive attack or to fight bad current design (http by default on naked fqdn) called to change in a near future.

By the way, in a post-Snowden era and given the current (un)security state of the Internet, a decent CA must at least **allow** sane people to do sane things and so must allow https only setup. Currently, LE **forbid** sane people to do things they consider more secure and is the future path of the Internet.

And worse, there is currently **no** decent justification of why LE disallow HTTPS for HTTP-01 challenge. Arguing about 80 port open or closed would never have been the primary discussion or is just useless if there is no good justification for this 443 port forbidden.

Reply:
I think you're still incorrectly threat modeling. The only way a passive listener can observe cookies is if the website failed to set the secure flag on them. Are you suggesting that websites which set non-secure cookies can make themselves safe by closing port 80? Because that seems like a very bad idea.

Reply:
I perfectly know what threat modeling is, but you seems to forget than bastion security expecting 1 level single barrier of protection is 70's era and current best practices are to have a in-depth security with each level assuming all other levels fucked up.
As a sysadmin, I don't want to be sure (and to have to test for) all applications are developped correctly and support secured cookie only. Worse, I must assume there is at least one badly configured, even by mistake during a recent deployment. And so I must proactively close the 80 port.

And yep, there is dumb website (and worse, banking website) configured with unsecured cookie and 80 port open, and worse without HSTS…

```
$ curl -sIL http://espace-assure.gmf.fr | rg -i "set-cookie|strict-transport"
Set-Cookie: BIGipServer-SA-gmf.fr=336407864.20480.0000;domain=.gmf.fr; path=/
$ curl -sIL https://espace-assure.gmf.fr | rg -i "set-cookie|strict-transport"
Set-Cookie: BIGipServer-SA-gmf.fr=336407864.20480.0000;domain=.gmf.fr; path=/
```

In this typical case, sysadmin can eventually immediately close the 80 port and expect a 
software (surely COBOL writen on a punch card and hosted on an AS.400 IBM mainframe…) change before the next millenium.

(And again, we eventually have to discuss 80 port opening only when we will agree about 443 port closing for HTTP-01. If there is no good reason to forbid 443 port for ACME, there is no reason to forbid 443 only setup and so no good reason to have 80 port still opened.)

Reply:
If you're deploying software that sets non-secure cookies, you are better off mitigating that with a reverse proxy that injects the secure flag into all cookies set by your application. Closing port 80 does not actually accomplish your goal of securing a non-secure application, because your users will still be vulnerable to an sslstrip-style attack. I do think we're honing in on the source of our disagreement here: It sounds like you consider a passive attack significantly more likely than an active one. In my opinion, the capabilities involved are very similar, and it's generally best to model all network attackers as active rather than deploy half measures that only work if your adversary chooses to limit themselves to passive attacks. It sounds like you disagree, and we may have to leave it at that.

Reply:
Yep, I disagree because Snowden prove exactly the opposite.
Passive SIGINT is far more deployed and used than active attack. The too famous NSA "collect it all"…

And if you are under active attack, you are the direct target (or somebody around you), and attacker is usually closed to you. So you are generally doomed and nothing can easily be done to mitigate the attack.